### PR TITLE
Lock markers and groups

### PR DESCRIPTION
**Pull request by radoering**
# Pull Request Check List

Requires: #9345
Requires: python-poetry/poetry-core#733

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

## First commit

Write information about the resulting groups that require a locked package and the resulting marker to the lock file. (Information is only written, not used yet.)

## Second commit
Introduce a new config option `installer.re-resolve` (default `true`); when set to `false` the marker information from the lock file is used, there is no re-resolving anymore when installing from the lock file.

# Motivation

What triggered me was the discussion at https://discuss.python.org/t/lock-files-again-but-this-time-w-sdists and especially https://github.com/brettcannon/peps/blob/lock-file/peps/pep-9999.rst, which might become a draft for a general lock file. To support this lock file format, we have to lock the resulting markers for packages so that re-resolving is not required for installation.

Further
- We can simplifiy the dependency walker in `poetry-plugin-export` and get rid of `dependency walk failed` issues.
  python-poetry/poetry-plugin-export#286
- We can do other stuff like creating an extra with pinned depenencies (not sure if that is something we want or if it is more something for a plugin).
  #9428

# Performance

We have to distinguish between locking and installing.

## Locking

There is an additional step to calculate the resulting markers so this change can only make things slower. However, I could not measure a significant increase in time required for locking for several projects. I suppose that's because what we are doing in the additional step are mostly marker operations, which have probably been cached during dependency resolution before.

## Installing

Evaluating markers instead of re-resolving could be faster or slower. However, I also could not measure a significant effect here. I suppose determining the exact packages to install is just not a bottleneck compared to the actual installation. Even when all packages were already installed (and installation is a noop), I could not measure a difference. However, there is a side effect:

### Side effect on installation process

With re-resolving we are installing batches of packages in parallel. The batches are determined by the depth in the dependency graph (same depth -> same batch). Since we do not lock the depth, without re-resolving all packages are in the same batch.

On the one hand, this can have a positive effect on performance because we do not have to wait for the last package of a batch.
(I could observe the complete installation of the dependencies of a largish project with all packages in the cache going down from 70s to 60s.)

On the other hand, this could have a negative effect on robustness (due to race conditions). I suppose, creating batches of packages with the same depth results in only installing packages in parallel that are (more or less) independent of each other.
(If a package depends on another package, they will normally have a different depth.) Parallel installation might be an issue if two packages write to the same directory (name space packages) and try to create it at the same time.

# Conclusion

Locking the resulting groups and markers opens up some new possibilities, especially for `poetry-plugin-export`. Installation without re-resolving is not a game changer on its own. Only the side effect will probably decrease the time for installation in some cases.

For now, I decided to make installation without re-resolving opt-in. We can also make it opt-out if we think it is safe enough and want to get more feedback.

### CODE REVIEW

**Review by Secrus — COMMENTED** [review]
Not my area of expertise, so not 100% sure, but looks pretty good to me, minor comments

File: src/poetry/installation/installer.py [hunk 1]
```diff
@@ -249,6 +258,13 @@ def _do_install(self) -> int:
                     "pyproject.toml changed significantly since poetry.lock was last"
                     " generated. Run `poetry lock [--no-update]` to fix the lock file."
                 )
+            if not reresolve and not self._locker.is_locked_groups_and_markers():
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/installer.py]: done

File: src/poetry/installation/installer.py [hunk 2]
```diff
@@ -259,46 +275,63 @@ def _do_install(self) -> int:
...
-            lockfile_repo = locked_repository
-
-        if not self.executor.enabled:
-            # If we are only in lock mode, no need to go any further
-            self._write_lock_file(lockfile_repo)
-            return 0
+            if reresolve:
+                lockfile_repo = locked_repository
+            else:
+                solved_packages = self._locker.locked_packages()
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/installer.py:282]: Different types: `locked_repository` is of type `LockfileRepository` and `locked_packages()` is of type `dict[Package, TransitivePackageInfo]`

File: src/poetry/installation/installer.py [hunk 3]
```diff
@@ -236,10 +233,18 @@ def _do_install(self) -> int:
             with solver.provider.use_source_root(
                 source_root=self._env.path.joinpath("src")
             ):
-                ops = solver.solve(use_latest=self._whitelist).calculate_operations()
+                solution = solver.solve(use_latest=self._whitelist)
+                solved_packages = solution.get_solved_packages()
```
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/installation/installer.py]: One more variable but one less line. No strong opinion so I changed it.

File: src/poetry/installation/installer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Secrus) [reply src/poetry/installation/installer.py:282]: ah, my eyes lied to me, didn't notice `locked_packages` vs `locked_repository`

**Review by Secrus — APPROVED** [review]

**Pull request merged**